File: examples/xml_import.cpp
#include <sys-sage.hpp>
#include <map>
#include <string>

using std::cout;
using std::endl;

using namespace sys_sage;


void usage(char* argv0)
{
    std::cerr << "usage: " << argv0 << " <hwloc xml path> <caps-numa-benchmark csv path>" << std::endl;
    std::cerr << "       or" << std::endl;
    std::cerr << "       " << argv0 << " (uses predefined paths which may be incorrect.)" << std::endl;
    return;
}

int main(int argc, char *argv[]) 
{
    std::string xmlPath;
    std::string bwPath;
    if(argc < 2){
        xmlPath = EXAMPLE_DIR + std::string("/skylake_hwloc.xml");
        bwPath = EXAMPLE_DIR + std::string("/skylake_caps_numa_benchmark.csv");
    }
    else if(argc == 3){
        xmlPath = argv[1];
        bwPath = argv[2];
    }
    else{
        usage(argv[0]);
        return 1;
    }

    // create root Topology and one node
    Topology *topo = new Topology();
    Node *n = new Node(topo, 1);

    cout << "-- Parsing Hwloc output from file " << xmlPath << endl;
    parseHwlocOutput(n, xmlPath);

    cout << "-- Parsing CapsNumaBenchmark output from file " << bwPath << endl;
    parseCapsNumaBenchmark((Component*)n, bwPath, ";");


    std::vector<Component *> c_orig = topo->FindDescendantsByType(ComponentType::Any);

    exportToXml(topo, "output.xml");

    Topology *topo2 = (Topology *)importFromXml("output.xml");

    for (auto c : c_orig) {
        int type = c->GetComponentType();
        int id = c->GetId();
        Component *c2 = topo2->GetDescendantById(id, type);
        if (c2 == NULL) {
            cout << "Component with id " << id << " and type " << type
                << " not found in imported topology" << endl;

            Component::DeleteSubtree(topo);
            Component::DeleteSubtree(topo2);

            // Delete output.xml
            //std::remove("output.xml");
            return 0;
        }
    }   
    cout << "Original topology:" << endl;
    topo->PrintSubtree();
    cout << "\nImported topology:" << endl;
    topo2->PrintSubtree();
    cout << "\nAll components found!" << endl;

    Component::DeleteSubtree(topo);
    Component::DeleteSubtree(topo2);

    // Delete output.xml
    std::remove("output.xml");
    return 0;
}

File: examples/xml_import.py
import py_sys_sage as pysage
import os
import sys

def main():
    xmlPath = ""
    if len(sys.argv) < 2:
        path_prefix = os.path.dirname(os.path.abspath(sys.argv[0]))
        xmlPath = os.path.join(path_prefix, "example_data", "skylake_hwloc.xml")
    elif len(sys.argv) == 3:
        xmlPath = sys.argv[1]

    # create root Topology and one node
    topo = pysage.Topology()
    n = pysage.Node(topo, 1)

    print("-- Parsing Hwloc output from file", xmlPath)
    pysage.parseHwlocOutput(n, xmlPath)

    c_orig = topo.GetComponentsInSubtree()

    pysage.exportToXml(topo, "output.xml")

    topo2 = pysage.importFromXml("output.xml")

    for c in c_orig:
        type = c.GetComponentType()
        id = c.GetId()
        c2 = topo2.GetSubcomponentById(id, type)
        if c2 is None:
            print(f"Component with id {id} and type {type} not found in imported topology")

            topo.Delete(True)
            topo2.Delete(True)

            # Delete output.xml
            os.remove("output.xml")
            return

    print("Original topology:")
    topo.PrintSubtree(2)
    print("\nImported topology:")
    topo2.PrintSubtree(2)
    print("\nAll components found!")

    topo.Delete(True)
    topo2.Delete(True)

    # Delete output.xml
    os.remove("output.xml")

if __name__ == "__main__":
    main()

